runtime/parachains/src/scheduler_polkadot/mod.rs

```diff
@@ -74,17 +72,19 @@ impl<T: Config> AssignmentProvider<T> for Pallet<T> {
 		}
 	}
 
-	fn push_assignment_for_core(core_idx: CoreIndex, assignment: Assignment) {
-		let parachain_cores = NumParachains::<T>::get()
-			.unwrap_or_else(|| <crate::scheduler_parachains::Pallet<T>>::session_core_count());
-		if (0..parachain_cores).contains(&core_idx.0) {
-			<crate::scheduler_parachains::Pallet<T>>::push_assignment_for_core(core_idx, assignment)
-		} else {
-			let _core_idx = CoreIndex(core_idx.0 - parachain_cores);
-			todo!()
-			//<crate::scheduler_parathreads::Pallet<T>>::push_assignment_for_core(
-			//	core_idx, assignment,
-			//)
+	fn push_parasentry_for_core(core_idx: CoreIndex, entry: ParasEntry) {
+		if entry.retries == 0 {
+			let parachain_cores = NumParachains::<T>::get()
```

Comment:
Why do we need this `NumParachains`?

Reply:
I see, for buffering across session boundaries, but is this even necessary? Can we not just flip the order of calls [here](https://github.com/paritytech/polkadot/blob/dbae30efe080a1d41fe54ef4da8af47614c9ca93/runtime/parachains/src/initializer.rs#L264)?

In any case, if we end up really needing it, `NumParachains` is a top level storage entry, it ought to have docs on what it does and why it is necessary. Also the name could be better, maybe something like `BufferedNumParachains` ... would at least give a hint on what it does.

Thinking more of it, I really don't think we should have it at all:

We do have an initialization sequence [here](https://github.com/paritytech/polkadot/blob/dbae30efe080a1d41fe54ef4da8af47614c9ca93/runtime/parachains/src/initializer.rs#L264) .. yet we have the scheduler tell the pars module about the session change [here](https://github.com/paritytech/polkadot/blob/d92b709979533d0b07d42e1548912becc8b21b6a/runtime/parachains/src/scheduler.rs#L156).

One of them should go. We either make paras internal to the scheduler - so no top-level entry any more, if that is sound or the other way round: Make the scheduler rely on the outside initializer to tell any assignment provider about new sessions. In that case order requirements should be documented. (Any assignment provider must be notified about a new session after the scheduler.)

Reply:
> Can we not just flip the order of calls [here](https://github.com/paritytech/polkadot/blob/dbae30efe080a1d41fe54ef4da8af47614c9ca93/runtime/parachains/src/initializer.rs#L264)?

Flipping them breaks a few `paras_inherent` unit tests. But also makes the adder collator test go from taking ~34 block to complete to ~42 blocks.

> Make the scheduler rely on the outside initializer to tell any assignment provider about new sessions.
Isn't this what we currently have? `initializer_on_new_session`  in the scheduler calls `AssignmentProvider::new_session`?

Reply:
> Isn't this what we currently have? `initializer_on_new_session` in the scheduler calls `AssignmentProvider::new_session`?

Precisely, while I suggest that with the flipped order that would be no longer necessary. Then it would be solely the initializer's business to tell the assignment providers about new sessions. Either that or make the assignment providers internal to the scheduler and not visible to the initializer.

Reply:
> But also makes the adder collator test go from taking ~34 block to complete to ~42 blocks.

Would be interesting to find out why.